### PR TITLE
fix(umap): only copy existing attrs

### PR DESCRIPTION
**Pull request by lmeyerov**
Some patches to make the program below run

* umap() only copies attrs that exist
* tests

----

### Original bug

in google colab:

```
# ! pip install --user graphistry[ai]
import pandas as pd
import graphistry
df = pd.DataFrame({'x': ['aa a' * 10, 'bb b' * 2, 'ccc '*20, 'dd abc', 'ee x1z'] * 10, 'y': [1, 2, 3, 4, 5] * 10})
graphistry.nodes(df).umap()
```

=>

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
[<ipython-input-21-6bceda2fa49e>](https://localhost:8080/#) in <cell line: 4>()
      2 import graphistry
      3 df = pd.DataFrame({'x': ['aa a' * 10, 'bb b' * 2, 'ccc '*20, 'dd abc', 'ee x1z'] * 10, 'y': [1, 2, 3, 4, 5] * 10})
----> 4 graphistry.nodes(df).umap()

1 frames
[~/.local/lib/python3.10/site-packages/graphistry/umap_utils.py](https://localhost:8080/#) in _process_umap(self, res, X_, y_, kind, memoize, featurize_kwargs, verbose, **umap_kwargs)
    401             fresh_res = copy.copy(res)
    402             for attr in ["_xy", "_weighted_edges_df", "_weighted_adjacency"]:
--> 403                 setattr(fresh_res, attr, getattr(old_res, attr))
    404             # have to set _raw_data attribute on umap?
    405             fresh_res._umap = old_res._umap  # this saves the day!

AttributeError: 'Plotter' object has no attribute '_weighted_adjacency'
```

---

### New bug

Fixing the above now leads to a new error, see new unit test

Note that I switched from `int` to `float` to make even easier for umap..

```
______________________ TestUMAPMethods.test_umap_simplest ______________________

self = <test_umap_utils.TestUMAPMethods testMethod=test_umap_simplest>

    @pytest.mark.skipif(not has_umap, reason="requires umap feature dependencies")
    def test_umap_simplest(self):
        df = pd.DataFrame({
            'x': ['aa a' * 10, 'bb b' * 2, 'ccc ' * 20, 'dd abc', 'ee x1z'] * 10,
            'y': [1.0, 2.0, 3.0, 4.0, 5.0] * 10
        })
>       graphistry.nodes(df).umap()

graphistry/tests/test_umap_utils.py:385: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
graphistry/umap_utils.py:623: in umap
    res = res._process_umap(
graphistry/umap_utils.py:415: in _process_umap
    emb = res._umap_fit_transform(X_, y_, verbose=verbose)
graphistry/umap_utils.py:305: in _umap_fit_transform
    self.umap_fit(X, y, verbose=verbose)
graphistry/umap_utils.py:288: in umap_fit
    self._umap.fit(X, y)
pygraphistry/lib/python3.8/site-packages/umap/umap_.py:2269: in fit
    X = check_array(X, dtype=np.float32, accept_sparse="csr", order="C")
pygraphistry/lib/python3.8/site-packages/sklearn/utils/validation.py:797: in check_array
    dtype_orig = np.result_type(*dtypes_orig)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (), kwargs = {}, relevant_args = ()

>   ???
E   ValueError: at least one array or dtype is required

<__array_function__ internals>:200: ValueError
------------------------------ Captured log call -------------------------------
WARNING  root:feature_utils.py:1087 DataFrame **seems** to be EmptyWARNING  graphistry.umap_utils:umap_utils.py:258 * Ignoring target column of shape (50, 0) in UMAP fit, as it is not one dimensional
```

### CODE REVIEW

**Comment by lmeyerov** [comment]
@silkspace ideas?

**Comment by lmeyerov** [comment]
adding a second numeric column `z` makes it run without exception

**Comment by dcolinmorgan** [comment]
thoughts:
- initially tried `remove_node_column=False` as i used a similar trick a while back, but no such luck
- your strings are just long enough for ngram featurization with spaces
- umap / everyone loves floats :)
- seems `X` string featurization isnt being input to umap, but your first fix patched that

seemingly simple solution:

- changing df column name to 'y' to anything -- guess we should reserve this var name
`df.rename(columns={'y':'yy'},inplace=True)`
- strangely 'X' doesn't have this problem, perhaps we just need to keep target placeholder free
- tested on redteam data too, same error changing any column to 'y' when only inputting 2 columns, 3+ columns gives no error (so then 'y' doesnt need placeholder to be free ??)

**Comment by lmeyerov** [comment]
tx --

re:colab, just to clarify that a fresh standard environment can hit this seemingly easily

re:changing y column name... Oh wow I totally missed that. I'm not sure if the original had a col with that name, that may have come in from my attempt to make a minimal reproduction. But maybe enough to close the immediate issue.

So sounds like something about 2 columns and/or y... maybe 1 col and no y has the same issue, it's about having a single col for X?

**Comment by dcolinmorgan** [comment]
interesting with 1 column
- changing name to anything other than 'x' seems to similarly rectify issue.
- then when you dont give the single column any name, and non-string 0 is placeholder returns an error expecting a string, so another tiny bug

```
def rename_small(df):
    if len(df.columns)==1:
        if (df.columns=='x') or (isinstance(df0.columns.to_list()[0],int)) or (df.columns=='y'):
            df.columns=['a']
    if len(df.columns)==2:
        if ('y' in df.columns.to_list()) or ('x' in df.columns.to_list()):
            df1=df.rename(columns={'x':'a','y':'b'},inplace=True)
    return df
```
obviously need to dig into more robust reason/solution

File: graphistry/feature_utils.py [hunk 1]
```diff
@@ -292,7 +292,14 @@ def remove_internal_namespace_if_present(df: pd.DataFrame):
         config.IMPLICIT_NODE_ID,
         "index",  # in umap, we add as reindex
     ]
-    df = df.drop(columns=reserved_namespace, errors="ignore")  # type: ignore
+
+    if (len(df.columns) <= 2):
```
Comment (lmeyerov) [inline graphistry/feature_utils.py:296]: great, can we get a `logger.debug` for this happening as it's not expected?

or better,

```python
class AutohealWarning(UserWarning)
  pass

...

  if len(df.columns) <= 2:
    warning.warn("...", AutohealWarning)
```

**Comment by lmeyerov** [comment]
Interesting `ci` fail:

```
INFO: pip is looking at multiple versions of flake8 to determine which version is compatible with other requirements. This could take a while.
Collecting flake8 (from graphistry==0+untagged.1.gd53a9d6)
  Downloading flake8-5.0.3-py2.py3-none-any.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.9/61.9 kB 22.5 MB/s eta 0:00:00
  Downloading flake8-5.0.2-py2.py3-none-any.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.9/61.9 kB 22.7 MB/s eta 0:00:00
  Downloading flake8-5.0.1-py2.py3-none-any.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 kB 22.5 MB/s eta 0:00:00
  Downloading flake8-5.0.0-py2.py3-none-any.whl (61 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.7/61.7 kB 22.8 MB/s eta 0:00:00
  Downloading flake8-4.0.1-py2.py3-none-any.whl (64 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.1/64.1 kB 23.0 MB/s eta 0:00:00
Collecting mccabe<0.7.0,>=0.6.0 (from flake8->graphistry==0+untagged.1.gd53a9d6)
  Downloading mccabe-0.6.1-py2.py3-none-any.whl (8.6 kB)
Collecting pycodestyle<2.9.0,>=2.8.0 (from flake8->graphistry==0+untagged.1.gd53a9d6)
  Downloading pycodestyle-2.8.0-py2.py3-none-any.whl (42 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.1/42.1 kB 15.7 MB/s eta 0:00:00
Collecting pyflakes<2.5.0,>=2.4.0 (from flake8->graphistry==0+untagged.1.gd53a9d6)
  Downloading pyflakes-2.4.0-py2.py3-none-any.whl (69 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 69.7/69.7 kB 27.1 MB/s eta 0:00:00
Collecting flake8 (from graphistry==0+untagged.1.gd53a9d6)
  Downloading flake8-4.0.0-py2.py3-none-any.whl (64 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 64.1/64.1 kB 24.1 MB/s eta 0:00:00
  Downloading flake8-3.9.2-py2.py3-none-any.whl (73 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.1/73.1 kB 26.3 MB/s eta 0:00:00
Collecting pyflakes<2.4.0,>=2.3.0 (from flake8->graphistry==0+untagged.1.gd53a9d6)
  Downloading pyflakes-2.3.1-py2.py3-none-any.whl (68 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 68.8/68.8 kB 23.1 MB/s eta 0:00:00
Collecting pycodestyle<2.8.0,>=2.7.0 (from flake8->graphistry==0+untagged.1.gd53a9d6)
  Downloading pycodestyle-2.7.0-py2.py3-none-any.whl (41 kB)
```

**Comment by lmeyerov** [comment]
ok line fail solved via https://github.com/graphistry/pygraphistry/pull/505

**Comment by lmeyerov** [comment]
tests went green!

**Pull request merged**